R/derive_locf_records.R

```diff
@@ -0,0 +1,184 @@
+#'   "SYSBP", "Systolic Blood Pressure (mmHg)",  4, "WEEK 4",
+#'   "SYSBP", "Systolic Blood Pressure (mmHg)",  6, "WEEK 6"
+#' )
+#'
+#' # Add a new record for each 'by_vars' when the input dataset does not contain
+#' # observations for missed visits/timepoints or when AVAL is missing.
+#' # Set DTYPE = 'LOCF' for the newly added records.
+#' # Carry forward the AVAL of the previous observation after sorting the dataset
+#' # by 'order' parameter.
+#' Final_advs <- derive_locf_records(data=advs,
+#'                                   dataset_expected_obs=advs_expected_obsv,
+#'                                   by_vars=vars(STUDYID, USUBJID, PARAM, PARAMCD),
+#'                                   order=vars(STUDYID, USUBJID, PARAM, PARAMCD, AVISITN, AVISIT))
+
+
+
+derive_locf_records <- function(dataset,
+                                dataset_expected_obs,
+                                by_vars,
+                                order) {
+  library(tidyr)
+
+  #### Input Checking ####
+
+  # Check by_vars and order variables in input dataset
+  assert_data_frame(dataset, required_vars = by_vars)
+  assert_data_frame(dataset, required_vars = order)
+
+  # Check if input parameters is a valid list of variables
+  assert_vars(by_vars, optional = TRUE)
+  assert_order_vars(order)
+
+
+
+  #### Prepping 'dataset_expected_obs' ####
+
+  # Add temp_dtype='LOCF' for the 'expected observation dataset'
+  dataset_expected_obs <- dataset_expected_obs %>%
+    mutate(temp_dtype="LOCF")
+
+
+  # Get the STUDYID and USUBJID from input dataset to set expected
+  # observations for each subject
+  subjids <- dataset %>%
+    select(STUDYID, USUBJID) %>%
+    distinct()
+
+  exp_obsv <- subjids %>%
+    crossing(dataset_expected_obs)
+
+
+
+  ##### Add LOCF records ####
+
+  # Split input dataset into the missing and non-missing AVAL records
+  aval_missing <- dataset %>%
+    filter(is.na(AVAL))
+
+  aval_not_missing <- dataset %>%
+    drop_na(AVAL)
+
+
+  # Get the variable names from the expected observation dataset
+  exp_obs_vars <- exp_obsv %>%
+    select(-temp_dtype) %>%
+    colnames()
+
+
+  # Get unique combination of visits/timepoints per parameter per subject
+  # from the original input dataset (with non-missing AVAL)
+  advs_unique_original <- aval_not_missing %>%
+    select(all_of(exp_obs_vars)) %>%
+    drop_na() %>%
+    distinct()
+
+
+  # Get all the expected observations that are to be added to the input
+  # dataset (with non-missing AVAL)
+  advs_exp_obsv3 <- exp_obsv %>%
+    anti_join(advs_unique_original, by=c(exp_obs_vars))
+
+
+  # Merge the expected observations with the input dataset (with non-missing AVAL)
+  # Arrange the dataset by 'order' and group it by 'by_vars'
+  # Use fill() to fill the AVAL from the previous observation for the newly added records
+  aval_not_missing_locf <- aval_not_missing %>%
+    full_join(advs_exp_obsv3,  by=c(exp_obs_vars)) %>%
+    mutate(DTYPE = case_when(temp_dtype=="LOCF" ~ "LOCF")) %>%
```

Comment:
Maybe it is better to use
```
mutate(DTYPE = if_else(temp_dtype == "LOCF", "LOCF", DTYPE, missing = DTYPE))
```
Then we do not overwrite `DTYPE` for the records of the input dataset in case `DTYPE` is already in the input dataset.

Reply:
This would throw an error if `DTYPE` is not already present in the dataset.
So, I've added an additional condition to check if `DTYPE` exists in the input dataset and then assign `DTYPE` accordingly.